### PR TITLE
`index` event

### PR DESCRIPTION
**Pull request by piranna**
Allow to render a custom `index` page from a function.

### CODE REVIEW

**Comment by piranna** [comment]
Any update on this?

File: index.js [hunk 1]
```diff
@@ -397,8 +395,8 @@ SendStream.prototype.isRangeFresh = function isRangeFresh(){
  */
 
 SendStream.prototype.redirect = function redirect(path) {
-  if (listenerCount(this, 'directory') !== 0) {
-    this.emit('directory')
+  if (listenerCount(this, 'directory')) {
+    this.emit('directory', path)
```
Comment (dougwilson) [inline index.js]: Adding a new argument to the `directory` event does not seem related to adding an `index` event. Can you please restructure this pull request to only do what was described in the pul request summary and ping me when you have updated it?

**Comment by piranna** [comment]
I've just rebased my changes over master and tests are passing, would you be able to merge it?

**Comment by dougwilson** [comment]
The comment I made on your code hasn't been addressed yet. Please address it and I'll take another look :)!

**Comment by piranna** [comment]
> The comment I made on your code hasn't been addressed yet. Please address it and I'll take another look :)!

Sorry, I missed the comment :-P I've just fixed it and move the `path` argument change on a [different pull-request](https://github.com/pillarjs/send/pull/115) :-)

**Comment by piranna** [comment]
Code rebased and tests passing. Can we focus on that over the other pull-requests? This is the most important for NodeOS at this moment...

**Comment by dougwilson** [comment]
Sure! Can you use typeof instead of instanceof for function checks? Why do we need both an option and an event? Is it not possible to just have one or the other?

**Comment by piranna** [comment]
At first I added an event like 'directory' and so, but later found it was
easier to use an optional callback. In fact, if I would redesign the API, I
would remove all the events except error and just use optional callbacks,
it doesn't makes sense to be able to register several functions that can
return a value for a single input (the path)...

**Comment by piranna** [comment]
I have just changed `instanceof` for `typeof`.

**Comment by dougwilson** [comment]
Cool! Why do we need both an option and an event? Is it not possible to just have one or the other? If we can use events over function options that would better fit in the current API. I know the API may not be what everyone is used to, but it is what I have inherited, and it doesn't sound like you want to wait for a discussion on changing the way the API works to land this feature :) At the very least, you need to move the test you made for the index event out of the section for the index option and add documentation for that to the README as well.

File: index.js [hunk 2]
```diff
@@ -142,13 +142,20 @@ function SendStream (req, path, options) {
     this._dotfiles = undefined
   }
 
-  this._extensions = opts.extensions !== undefined
-    ? normalizeList(opts.extensions, 'extensions option')
```
Comment (dougwilson) [inline index.js]: Is this a necessary change for this PR?

File: README.md [hunk 3]
```diff
@@ -229,7 +234,7 @@ var app = http.createServer(function onRequest (req, res) {
 }).listen(3000)
 ```
 
-## License
```
Comment (dougwilson) [inline README.md]: Is this a necessary change for this PR?

File: test/send.js [hunk 4]
```diff
@@ -702,9 +702,11 @@ describe('send(file).pipe(res)', function () {
         .pipe(res)
       })
 
+      var filePath = path.join(fixtures, 'pets', 'index.html')
+
       request(app)
       .get('/pets/')
-      .expect(200, fs.readFileSync(path.join(fixtures, 'pets', 'index.html'), 'utf8'), done)
+      .expect(200, fs.readFileSync(filePath, 'utf8'), done)
```
Comment (dougwilson) [inline test/send.js]: Is this a necessary change for this PR?

File: index.js [hunk 5]
```diff
@@ -542,11 +549,16 @@ SendStream.prototype.pipe = function pipe (res) {
   }
 
   // index file support
+  if (typeof this._index === 'function') {
```
Comment (dougwilson) [inline index.js]: Can this check just be moved into the sendIndex method? It is a public method, and it's weird the event emit is in there but not this check.

File: README.md [hunk 6]
```diff
@@ -81,9 +81,14 @@ This is skipped if the requested file already has an extension.
 
 ##### index
 
-By default send supports "index.html" files, to disable this
-set `false` or to supply a new index pass a string or an array
-in preferred order.
+By default *send* supports "index.html" files. To disable this, set index option
```
Comment (dougwilson) [inline README.md]: Since this section is about the index option, I don't think you need to refer to itself, as it ends up being really wordy.

File: README.md [hunk 7]
```diff
@@ -81,9 +81,14 @@ This is skipped if the requested file already has an extension.
...
+to `false`, or you can be able to supply a new index passing a string with the
```
Comment (dougwilson) [inline README.md]: "you can be able to" doesn't sound right. I think probably just remove those words?

Comment (dougwilson) [inline README.md]: And passing would become just pass

File: README.md [hunk 8]
```diff
@@ -81,9 +81,14 @@ This is skipped if the requested file already has an extension.
...
+filename or an array in the preferred order.
+
+Alternatively, you can pass a function that will generate the index. The
```
Comment (dougwilson) [inline README.md]: Rather than tacking this on as a separate paragraph, can you incorporate them all together? Otherwise, it sounds very hack in, but having a complete paragraph saying that the option only takes a boolean, string, or array, then another paragraph following to say "but wait! We forgot to add it takes a function too"

File: README.md [hunk 9]
```diff
@@ -81,9 +81,14 @@ This is skipped if the requested file already has an extension.
...
+function will receive the filesystem absolute path and will have the
+`SendStream` instance as `this` object. You should call internally to the
```
Comment (dougwilson) [inline README.md]: Having the "this" populated on a callback like function is very odd, and very easy to get lost. It is kind of annoying already for events, but typically I never see this on straight functions. Can everything a user would need just be passed as arguments?

File: README.md [hunk 10]
```diff
@@ -81,9 +81,14 @@ This is skipped if the requested file already has an extension.
...
+`this.send()` method or try to mimic its behaviour.
```
Comment (dougwilson) [inline README.md]: Can the docs be clearer on how someone is supposed to "mimic its behavior"?

File: README.md (hunk 7 above)
Comment (piranna) [inline README.md]: That's formal english... As you want.

**Comment by dougwilson** [comment]
And I was jus able to sit down a s review your pull request, woot! We can get this moving along for sure :) I just had a few questions and comments; some in this conversation itself and some inline in the code.

File: README.md (hunk 3 above)
Comment (piranna) [inline README.md]: This clean-up is done automatically by Atom. Correct markdown don't accept blanks at the end of the line, except two blank spaces to force to do backline.

File: index.js (hunk 2 above)
Comment (piranna) [inline index.js]: `normalizeList()` now return an empty list when the input is `undefined`, so the check here is bloated.

File: index.js (hunk 5 above)
Comment (piranna) [inline index.js]: I would in fact remove the `index` event at all... This would be moved there too, but then the check for the function would need to be added both on the next `if` conditional AND inside the `sendIndex()` function too.

File: test/send.js (hunk 4 above)
Comment (piranna) [inline test/send.js]: ESlint was giving me errors of lines too long, this fixed it. Source code lines should never be longer than 80 columns.

File: index.js (hunk 2 above)
Comment (dougwilson) [inline index.js]: I don't see any changes to  normalizeList  in your PR. Perhaps I should be clearer: don't include changes in your PR that are not accomplishing your goal. These types of changes need to be different PRs. Please revert this change.

File: test/send.js (hunk 4 above)
Comment (dougwilson) [inline test/send.js]: Revert the change if you would like this landed... Please don't include unrelated changes. If we wanted to enforce a max line length, we would do so, but we do not.

File: index.js (hunk 5 above)
Comment (dougwilson) [inline index.js]: I'm sorry, I don't understand your response. Can you say it in a different way?

File: README.md (hunk 7 above)
Comment (dougwilson) [inline README.md]: > That's formal english... As you want.

I'm sorry, I can't seem to figure out what your response means...

**Comment by piranna** [comment]
> Cool! Why do we need both an option and an event? Is it not possible to just have one or the other?

Yes, we could use just one of them, and in fact it would be better to use optional callbacks only, I just added the event to be on pair with the API, but I would remove them.


> If we can use events over function options that would better fit in the current API. I know the API may not be what everyone is used to, but it is what I have inherited, and it doesn't sound like you want to wait for a discussion on changing the way the API works to land this feature :)

I'm open for a discussion about changing how the API works, it's just that this seemed to be ready to merge and wanted to use upstream code ASAP. I think events-based API should be deprecated and use optional callbacks instead, it's the correct behaviour and what everybody expects:

* just one and only one listener: callback
* zero or more listeners
  * just one and only one call: promise
  * zero or more calls: event

This also makes the API more explicit about what can and what can't do? What do you think?


> At the very least, you need to move the test you made for the index event out of the section for the index option and add documentation for that to the README as well.

Where should I put the test for the `index` event in case we still use it?

**Comment by dougwilson** [comment]
We can discuss a new API style for this module, absolutely, but no new API style will be merged for months to come, because this module is locked to the release schedule of Express.js itself. Even this new feature will not get released until a minor release of Express.Js is scheduled. I wish we had all this conversation 3 months ago,when I started commenting on the changes, as it would have ended up with the 4.14 Express.js release.

I don't have time this weekend to discuss and think about an entire change to the API of this module, only have been able to sneak in reviewing these changes as a simple feature addition, as it sounded like you were interested in landing. I'll try to circle back to this PR another day when I have some time to discuss a new API :)

**Comment by piranna** [comment]
> this module is locked to the release schedule of Express.js itself

That's totally insane, being so much specific it should have its own lifecycle, specially for this backward compatible things... And if not, semantic versioning is just for that...


> I'll try to circle back to this PR another day when I have some time to discuss a new API :)

Ok, send me an email when you have time to discuss it.

**Comment by dougwilson** [comment]
> That's totally insane, being so much specific it should have its own lifecycle

You're always welcome to use another module or fork this module. This is not up for discussion, I'm sorry. This code could have easily just been embedded into Express core. It is broken out for you to use it without Express, not to have a different release schedule, I'm sorry this is an inconvenience to you. We barely have enough people to maintain all this module, let alone for them all yo have different release timelines,meaning we have to create multiple lineages for modules when security vulnerabilities come out.

I feel like this is getting a little too headed foe me, so I'm not longer going to participate in this PR for time time being. Another member of the organization is welcome to pick up moving this PR forward...

**Comment by piranna** [comment]
I'm currently using a fork of this repo, the point of willing my changes to be merged is to clean up and have a consistent repo, but just by suggesting to use a repo, although acceptable, only contribute to degrade npm ecosystem. On NodeOS we lack a lot of people too, being so picky with the pull-requests just makes everybody losing their time and willing to not contribute anymore. Everybody lose :-(

**Comment by dougwilson** [comment]
Your PR is just never going to land, as you have not even hinted that you are willing to make any changes to your PR, and our philosophies for maintain seems to be too different: You suggest that I just merge anything without question, and I only want to merge quality code. These are unfortunately mutually exclusive. If you want to get a feature added, please work to add the feature with the API we have, rather than debate which API style is better.

**Comment by piranna** [comment]
I did all the changes that you tell me to do. You are not looking for quality code, but a clean and verbose commits history. If you really wanted to have quality code, you'll start by thinking about changing the API, but I only hinted about that and tried to adapt to the current code style and project organization. If you are not willing to maintain this project is ok, just give me permissions and I'll maintain it in your place.

**Comment by dougwilson** [comment]
> I did all the changes that you tell me to do.

I'm not sure why you say this, as there are many, many open comments on your code above, and no commits since any of them. No changes I can see.

> just give me permissions and I'll maintain it in your place

No.
